### ISSUE TITLE
Designing a "multi-tenant" CoAP server

### ISSUE THREAD

**Issue by beriberikix** [obgm/libcoap#351]
Hello,

I'm designing a CoAP server intended to support multiple users, or multi-tenant, and would love some suggestions on how to improve the design's use of `libcoap`. I'm relatively new to UNIX networking and libcoap, so I'm using this as a longish learning exercise, so expect novice-level questions!

When I say "multi-tenant" I mean that each user should have their own subdomain and namespace for developing request handlers. For example, `coap://foo.example.com/bar` and `coap://baz.example.com/bar` should respond differently because the owner of `foo` and `baz` each develop their handlers separately. A different service will be managing the registration and management of the subdomains.

I have a proof of concept that I'm hacking on over [here](https://github.com/coapcloud/coapc_fe) (very much hacky at the moment.) It has a basic GET handler that does two basic features:

1. Parses the `URI-Host` for the subdomain
2. Attaches response details via the resource User Data

Note that I haven't pushed the full implementation to the repo yet. With that information, the handler can reason, "for subdomain `foo`, respond with x based on the user data."

In the future I plan on adding:

1. An endpoint to register and deregister resources at runtime
2. The ability for handlers to use the User Data to call out to an RPC to honor the request. The RPCs will map to each user's business logic for their resources.
3. A load balancer like nginx to support scaling up

Anyone have suggestions on how to improve the design, for scalability and/or maintainability, are welcome. I was originally hoping to make a more stateless system, perhaps running on Kubernetes with a local cache of the RPC logic in Redis but have tabled that due to my own level of expertise. I'm concerned that my naive understanding won't be memory efficient or scalable. For example, should I be initializing resources and/or contexts differently? Are there hooks in `libcoap` I should look into extending in order to more performantly handle this use case?

Thank you in advance!

**Comment by mrdeep1** [comment]
@beriberikix 

> An endpoint to register and deregister resources at runtime

An endpoint is listening for traffic on a specific port with a defined protocol (UDP/TCP).  Unless you are planning to use different ports, this is not the way to move forward.  Resources can be added to / removed from a CoAP context.  See coap_resource(3) in the documentation.
[Only a single CoAP context should be needed]

> A load balancer like nginx to support scaling up

You need to consider using epoll support for this, but need PR #347 merged into the develop branch to do this.  The code within the libcoap library is only single threaded (i.e multiple threads cannot concurrently work in parallel that are doing coap_*() function calls).

**Comment by beriberikix** [comment]
Thanks for the feedback, @mrdeep1!

>  Unless you are planning to use different ports, this is not the way to move forward.

Can you expand on what you mean? If I pass a shared Context into the function that registers new Resource at runtime and call `coap_add_resource`, what issues might I run into?

RE: epoll - that would be great! I'm assuming that means the implementation would be scoped to just UNIX systems? Not that I'm targeting Windows but libcoap seems to have support. Have you looked @ https://github.com/libuv/libuv? I know it is a x-platform abstraction used by [many projects](https://github.com/libuv/libuv/wiki/Projects-that-use-libuv), including Node.js.

**Comment by mrdeep1** [comment]
> Unless you are planning to use different ports, this is not the way to move forward.
> 
> Can you expand on what you mean? If I pass a shared Context into the function that registers new Resource at runtime and call coap_add_resource, what issues might I run into?

The context of this (different ports) was endpoints, not resources.  There should be no issues with coap_add_resource().

epoll support is only enabled if the underlying O/S supports it (yes, typically Linux). If running under Windows or some embedded O/S, then the support will not be there unfortunately.

No, I have not looked at https://github.com/libuv/libuv , so do not know whether it will help here or not, and it is unlikely that I would have time to get libcoap to work with it.

**Comment by mrdeep1** [comment]
@beriberikix libuv is not likely to get into libcoap anytime soon unless a contributor takes on the responsibility to do it.  Otherwise, epoll support is available in hte develop branch.  Can this be closed off now?
